# HG changeset patch
# Parent  fb54b333744954208741b14c89100f273e1b63db
Properly walk over json objects and arrays

diff --git a/core/jsonrpc.c b/core/jsonrpc.c
--- a/core/jsonrpc.c
+++ b/core/jsonrpc.c
@@ -295,19 +295,20 @@
                                      int parsed,
                                      const char *json)
 {
-    assert(tokens[ti].type == JSMN_OBJECT || tokens[ti].type == JSMN_ARRAY);
+    assert(lgtd_jsonrpc_type_object_or_array(&tokens[ti], json));
     assert(ti < parsed);
 
-    int objsize = tokens[ti].size;
-    if (tokens[ti++].type == JSMN_OBJECT) {
-        ti++; // move to the value
-    }
+    jsmntype_t container_type = tokens[ti].type;
+    int objsize = tokens[ti++].size;
     while (objsize-- && ti < parsed) {
-        if (tokens[ti].type == JSMN_OBJECT || tokens[ti].type == JSMN_ARRAY) {
+        if (container_type == JSMN_OBJECT) {
+            ti++; // move to the value
+        }
+        if (lgtd_jsonrpc_type_object_or_array(&tokens[ti], json)) {
             ti = lgtd_jsonrpc_consume_object_or_array(tokens, ti, parsed, json);
-       } else {
-            ti += tokens[ti].size + 1; // move to the next value
-       }
+        } else {
+            ti++;
+        }
     }
 
     return ti;
diff --git a/tests/core/jsonrpc/test_jsonrpc_consume_object_or_array.c b/tests/core/jsonrpc/test_jsonrpc_consume_object_or_array.c
new file mode 100644
--- /dev/null
+++ b/tests/core/jsonrpc/test_jsonrpc_consume_object_or_array.c
@@ -0,0 +1,34 @@
+#include "jsonrpc.c"
+
+#include "mock_client_buf.h"
+#include "test_jsonrpc_utils.h"
+
+int
+main(void)
+{
+    jsmntok_t tokens[32];
+    memset(tokens, 0, sizeof(tokens));
+    const char json[] = ("["
+        "{"
+            "\"method\": \"power_on\","
+            "\"id\": \"004daf12-0561-4fbc-bfdb-bfe69cfbf4b5\","
+            "\"params\": [\"*\"], \"jsonrpc\": \"2.0\""
+        "},"
+        "{"
+            "\"method\": \"get_light_state\","
+            "\"id\": \"1f7a32c8-6741-4ee7-bec1-8431c7d514dc\","
+            "\"params\": [\"*\"],"
+            "\"jsonrpc\": \"2.0\""
+        "}"
+    "]");
+    int parsed = parse_json(
+        tokens, LGTD_ARRAY_SIZE(tokens), json, sizeof(json)
+    );
+
+    int ti = lgtd_jsonrpc_consume_object_or_array(tokens, 0, parsed, json);
+    if (ti != parsed) {
+        errx(1, "ti %d (expected %d)", ti, parsed);
+    }
+
+    return 0;
+}
